### PR TITLE
Clean up block chain pkg initial sync

### LINKED ISSUE #6072: Refactor blockchain service package for clarity and add better coverage
# 💎 Issue

Blockchain service package is a mess. Opening this issue to track improvement plan. Todo:
* Remove `ReceiveBlockNoPubsubForkchoice`.  Fork choice is no longer an expensive process we can remove this interface for simplicity
* Group similar functions together in `ReceiveBlock` path
* Make functions live next to the usages. There's currently `receive_block.go` and `process_block.go` but that should just be one
* Rename functions for clarity
* Better comments for the functions
* Add better test coverage

Some reference docs:
Screen Shot 2020-06-01 at 11 12 46 AM

### PR DESCRIPTION
**Pull request by terencechain**
**What type of PR is this?**
> Cleanup

**What does this PR do? Why is it needed?**
Cleaned up initial sync (not batch) functions in block chain pkg. Change list:
* Inside `saveHeadNoDB` return early if the head root is the same as cached head root
* Add function `savePostStateInfo` which saves block, state and fork choice. Reuse `savePostStateInfo` for `onBlock` and `onBlockInitialSync`
* Use `handleEpochBoundary` for `onBlock`
* Move `MigrateToCold` to inside `updateFinalized`
* Invalidate hot state cache within `StateByRootInitialSync` to avoid setter `DeleteHotStateInCache`

**Which issues(s) does this PR fix?**

Fixes #6072 

**Other notes for review**
Tested run time

### CODE REVIEW

File: beacon-chain/blockchain/process_block.go [hunk 1]
```diff
@@ -250,51 +238,6 @@ func (s *Service) onBlockBatch(ctx context.Context, blks []*ethpb.SignedBeaconBl
...
-	if uint64(len(s.getInitSyncBlocks())) > initialSyncBlockCacheSize {
-		if err := s.beaconDB.SaveBlocks(ctx, s.getInitSyncBlocks()); err != nil {
-			return err
-		}
-		s.clearInitSyncBlocks()
-	}
-
-	// Update finalized check point. Prune the block cache and helper caches on every new finalized epoch.
-	if postState.FinalizedCheckpointEpoch() > s.finalizedCheckpt.Epoch {
-		if err := s.beaconDB.SaveBlocks(ctx, s.getInitSyncBlocks()); err != nil {
```
Comment (nisdas) [inline beacon-chain/blockchain/process_block.go:280]: dont we need this in the event of a finalized checkpoint being saved ?

File: beacon-chain/blockchain/process_block.go [hunk 2]
```diff
@@ -250,51 +238,6 @@ func (s *Service) onBlockBatch(ctx context.Context, blks []*ethpb.SignedBeaconBl
...
-			return err
-		}
-		s.clearInitSyncBlocks()
-
-		if err := s.updateFinalized(ctx, postState.FinalizedCheckpoint()); err != nil {
-			return err
-		}
-
-		fRoot := bytesutil.ToBytes32(postState.FinalizedCheckpoint().Root)
-		if err := s.stateGen.MigrateToCold(ctx, fRoot); err != nil {
```
Comment (nisdas) [inline beacon-chain/blockchain/process_block.go:290]: what is the reason for `migrateToCold` not being required now ?

File: beacon-chain/blockchain/process_block.go (hunk 1 above)
Comment (terencechain) [reply beacon-chain/blockchain/process_block.go:280]: We do this above. Saving blocks to DB every 2 epochs. It's not required to have these blocks in DB upon finalized check point. Tested run time as well.

```go
// Rate limit how many blocks (2 epochs worth of blocks) a node keeps in the memory.
	if uint64(len(s.getInitSyncBlocks())) > initialSyncBlockCacheSize {
		if err := s.beaconDB.SaveBlocks(ctx, s.getInitSyncBlocks()); err != nil {
			return err
		}
		s.clearInitSyncBlocks()
	}
```

File: beacon-chain/blockchain/process_block.go (hunk 2 above)
Comment (terencechain) [reply beacon-chain/blockchain/process_block.go:290]: It's still there, we moved it within `updatedFinalized`
https://github.com/prysmaticlabs/prysm/pull/6562/files#diff-f407600862245d51b706766ae42ee345R208

**Pull request merged**